code/parachain/frame/composable-maths/src/dex/constant_product.rs

```diff
@@ -123,6 +123,77 @@ where
 	Ok(ai)
 }
 
+/// Compute the amount of the input token given the amount of the output token.
+///
+/// If `Ok`, returns a tuple containing `(a_sent, fee)`.
+/// To get `a_sent` without accounting for the fee, set `f = 0`.
+/// Amount in, round up results.
+///
+/// **NOTE:** Weights must already be normalized.
+///
+/// # Parameters
+/// * `w_i` - Weight of the input token
+/// * `w_o` - Weight of the output token
+/// * `b_i` - Balance of the input token
+/// * `b_o` - Balance of the output token
+/// * `a_out` - Amount of the output token desired by the user
+/// * `f` - Total swap fee
+///
+/// From https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#41-fee-math-updates,
+/// equation (3)
+pub fn compute_in_given_out_new<T: PerThing>(
+	w_i: T,
+	w_o: T,
+	b_i: u128,
+	b_o: u128,
+	a_out: u128,
+	f: T,
+) -> Result<(u128, u128), DispatchError> {
+	ensure!(
+		a_out <= b_o,
+		DispatchError::from(
+			"`a_out` must not be graeter than `b_o` (can't take out more than what's available)"
```

Comment:
NIT: Consider a separate error type that implements `Into<DispatchError>` and `From<ArithmeticError>`, something like:

```rust
enum InGivenOutError {
    ArithmeticError(sp_runtime::ArithmeticError),
    CantTakeMoreThanIsAvailable,
}
```

This would make error testing a lot easier as well

Reply:
Good idea, having enumerable errors is way better than just messages.